counter_mspm0: isr callback is not thread safe

dkarnikis opened the pull request:
Check commits

DmtKats reviewed (commented):
I think this should be fine @hansbinderup I will wait to give you a chance to check before approving

hansbinderup reviewed (commented):
I'm mostly concerned about the sem actually being taken from the ISR. I think we need to use a `spin_lock` here instead. It will also make the rest of the interface ISR safe

hansbinderup commented on drivers/counter/counter_mspm0.c:
```diff
@@ -81,22 +78,20 @@ static int counter_mspm0_set_alarm(const struct device *dev, uint8_t chan,
+	K_SPINLOCK(&data->lock) {
 		struct counter_mspm0_channel_data *channel_data = &data->channel_data[chan];
 		if (channel_data->callback) {
 			LOG_ERR("Callback is already registered");
```
Please remove this log. We have the return value - should be fine :+1: 
FYI: `spin_lock` context is critical so only minimal work has to be performed here (as we just discussed). We can not guarantee that the log is offloaded why this might be problematic.

hansbinderup reviewed (approved):
Sweet, looks good :1st_place_medal:

dkarnikis commented:
@hansbinderup let me know if it is okay to merge

hansbinderup commented:
Maybe turn it into two commits? One for using the spin lock and one to correct IRQ configuration :)

dkarnikis-bo committed:
counter_mspm0: use spinlock instead of mutex

Remove mutex and use spinlock to avoid some locking issues. Due to the
fact that the spinlock disables the ISR while holding a lock, we can
totally remove any locking mechanism from the ISR context since it is
thread safe.

Signed-off-by: Dimitris Karnikis <dika@bang-olufsen.dk>

dkarnikis-bo committed:
counter_mspm0: set timer config on init

instead of setting the timer configuration, properties and IRQ on every
new command, just set them once on device driver init

Signed-off-by: Dimitris Karnikis <dika@bang-olufsen.dk>

hansbinderup commented on drivers/counter/counter_mspm0.c:
```diff
@@ -42,16 +42,7 @@ static int counter_mspm0_start(const struct device *dev)
-		DL_Timer_enablePower(cfg->timer);
-
-		DL_Timer_setClockConfig(cfg->timer, (DL_Timer_ClockConfig *)&cfg->clock_cfg);
 		DL_Timer_initTimerMode(cfg->timer, &data->time_cfg);
```
Shouldn't this be done in `counter_mspm0_set_alarm`?

dkarnikis replied:
you are correct, seems to do the trick

hansbinderup reviewed (approved):
Nice, now it looks pretty solid :)
